pilot/pkg/bootstrap/namespacecontroller.go

```diff
@@ -70,50 +114,63 @@ func NewNamespaceController(ca *ca.IstioCA, core corev1.CoreV1Interface) (*Names
 				return core.Namespaces().Watch(options)
 			}}
 	})
-	c.namespaceStore, c.namespaceController =
-		cache.NewInformer(namespaceLW, &v1.Namespace{}, namespaceResyncPeriod, cache.ResourceEventHandlerFuncs{
-			UpdateFunc: c.namespaceUpdated,
-			AddFunc:    c.namespaceAdded,
+	_, c.namespaceController =
+		cache.NewInformer(namespaceLW, &v1.Namespace{}, 0, cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.queue.Push(func() error {
+					return c.namespaceChange(newObj)
+				})
+			},
+			AddFunc: func(obj interface{}) {
+				c.queue.Push(func() error {
+					return c.namespaceChange(obj)
+				})
+			},
 		})
 	return c, nil
 }
 
-// When a namespace is created, Citadel adds its public CA certificate
-// to a well known configmap in the namespace.
-func (nc *NamespaceController) namespaceAdded(obj interface{}) {
+// Run starts the NamespaceController until a value is sent to stopCh.
+func (nc *NamespaceController) Run(stopCh <-chan struct{}) {
+	go nc.namespaceController.Run(stopCh)
+	go nc.configMapController.Run(stopCh)
+	cache.WaitForCacheSync(stopCh, nc.namespaceController.HasSynced, nc.configMapController.HasSynced)
+	log.Infof("Namespace controller started")
+	go nc.queue.Run(stopCh)
```

Comment:
nit: start the queue first

Reply:
What is the benefit? We don't seem to do it in other places: https://github.com/istio/istio/blob/216ce9b252ed0efc27fae6623f84880368e211fa/pilot/pkg/serviceregistry/kube/controller/controller.go#L363-L371

Reply:
I think this is fine as-is. We want to wait for the cache to be fully populated before we starting processing it. `go nc.queue.Run(stopCh)` is equivalent to starting workers in a typically k8s controller which happens after the sync (see 
https://github.com/kubernetes/sample-controller/blob/05e054c7a4a1296ad7a88e511adfd936dd7a2c3c/controller.go#L151-L175).

Reply:
Never mind. Start later can prevent unknown error caused by cache miss.